### PR TITLE
Full Pydantic Model

### PR DESCRIPTION
**Pull request by jsamoocha**
## Description

This replaces the complete legacy model with the extensions of the generated pydantic model. It adds the undocumented attributes of all legacy model classes.

Compared to previous iterations, handling nested entity (collections) is simplified by overriding the superclass attributes. The custom type extension validator is entirely removed.

Let's test this for a while and maybe then announce to the community the coming change and provide an opportunity to test from this `pydantic_model` branch before merging into master.

## Type of change

Refactoring

## Does your PR include tests

- [x] Yes

## Do all of the CI tests run successfully on this pr?

- [x] Check this when your PR has a green check.

### CODE REVIEW

File: stravalib/model.py [hunk 1]
```diff
@@ -417,32 +247,74 @@ class AthleteStats(
...
-
 
 class Athlete(
     DetailedAthlete, DeprecatedSerializableMixin, BackwardCompatibilityMixin, BoundClientEntity
 ):
-    is_authenticated: Optional[bool] = None
+    # field overrides from superclass for type extensions:
+    clubs: Optional[List[Club]] = None
+    bikes: Optional[List[Bike]] = None
+    shoes: Optional[List[Shoe]] = None
```
Comment (jsamoocha) [inline stravalib/model.py:273]: So it turns out that one can simply override a superclass field to extend its type. This allows simplification of the model and removal of the `extend_types()` custom validator factory.

File: stravalib/tests/unit/test_model.py [hunk 2]
```diff
@@ -48,28 +46,23 @@ def test_legacy_to_dict(self, model_class, attr, value):
...
-            None,
         ),
-        (Activity, {'timezone': 'Europe/Amsterdam'}, pytz.timezone('Europe/Amsterdam'), None)
+        (Activity, {'timezone': 'Europe/Amsterdam'}, pytz.timezone('Europe/Amsterdam'))
     ),
 )
-def test_backward_compatibility_mixin(
-    model_class, raw, expected_value, expected_warning
+def test_backward_compatibility_mixin_field_conversions(
+    model_class, raw, expected_value
```
Comment (jsamoocha) [inline stravalib/tests/unit/test_model.py:62]: Removed `expected_warning` param from the test: it was always None

File: stravalib/tests/unit/test_model.py [hunk 3]
```diff
@@ -88,6 +81,17 @@ def test_deserialization_edge_cases(model_class, raw, expected_value):
...
 
+def test_subscription_callback_field_names():
+    sub_callback_raw = {
+        "hub.mode": "subscribe",
+        "hub.verify_token": "STRAVA",
+        "hub.challenge": "15f7d1a91c1f40f8a748fd134752feb3",
+    }
+    sub_callback = SubscriptionCallback.parse_obj(sub_callback_raw)
+    assert sub_callback.hub_mode == 'subscribe'
+    assert sub_callback.hub_verify_token == 'STRAVA'
```
Comment (jsamoocha) [inline stravalib/tests/unit/test_model.py:92]: This is moved from an existing test to ensure the `.` in the field name is replaced with `_` to make a valid python identifier.

**Comment by jsamoocha** [comment]
It looks like the documentation generation we now use is not digesting the pydantic model fluently: the model documentation doesn't look nice. The following may help:

- https://sphinx-pydantic.readthedocs.io/en/latest/
- https://pypi.org/project/autodoc-pydantic/

**Comment by lwasser** [comment]
> * https://pypi.org/project/autodoc-pydantic/

No problem - if you want, i'll have a look at that and will see if i can get it to work more nicely as you've done so much work here . right now just pulling down the rest of the changes to test locally! then will build docs and see what's going on.

**Comment by lwasser** [comment]
@jsamoocha this is a ton of work but i'm also finding some quirks (as to be expected with a overhaul of this size). Do you want to merge this and we can open issues to handle specific things? here is one that i just found

```
response = []
act = client.get_activities(limit=4)
for anAct in act:
    for aActResponse in anAct:
        response.append(aActResponse)
```

the response object is odd as it has  both the bound client and then an athlete object that is full of "nones"

```
[('bound_client', <stravalib.client.Client at 0x158c044c0>),
 ('id', 8466714764),
 ('achievement_count', 0),
 ('athlete',
  Athlete(bound_client=None, id=10295934, city=None
```

i will keep playing just wondering if you want to merge this, then we can test?

File: docs/contributing/development-guide.md [hunk 4]
```diff
@@ -67,6 +67,26 @@ $ pip install -r requirements.txt
...
+## Architecture Overview
+
+![Stravalib Architecture](../images/stravalib_architecture.png)
+
+Stravalib contains the following main components:
+
+At the core, a (pydantic) domain model is generated and updated by a bot via pull requests.
+This model reflects the officially published API specification by Strava and is stored in
+the module `strava_model.py`. This file should never be edited other than by the pull requests
+generated by the bot.
```
Comment (lwasser) [inline docs/contributing/development-guide.md:79]:
```suggestion
the module `strava_model.py`. This file should never be edited  manually. Instead, the stravalib bot will suggest changes to the model through pull requests that can then be merged by stravalib maintainers.
```

File: docs/contributing/development-guide.md [hunk 5]
```diff
@@ -67,6 +67,26 @@ $ pip install -r requirements.txt
...
+
+For allowing custom typing, unit conversion, (de-)serialization behavior, and support of
+undocumented Strava features, the module `model.py` contains classes that inherit from the
+official Strava domain model in `strava_model.py`.
+
+The module `protocol.py` manages the low-level HTTP traffic (including rate limiting)
```
Comment (lwasser) [inline docs/contributing/development-guide.md:85]: what do you mean by low level in this case? could you clarify a bit?

File: docs/contributing/development-guide.md [hunk 6]
```diff
@@ -67,6 +67,26 @@ $ pip install -r requirements.txt
...
+Stravalib contains the following main components:
+
+At the core, a (pydantic) domain model is generated and updated by a bot via pull requests.
+This model reflects the officially published API specification by Strava and is stored in
+the module `strava_model.py`. This file should never be edited other than by the pull requests
+generated by the bot.
+
+For allowing custom typing, unit conversion, (de-)serialization behavior, and support of
+undocumented Strava features, the module `model.py` contains classes that inherit from the
+official Strava domain model in `strava_model.py`.
```
Comment (lwasser) [inline docs/contributing/development-guide.md:83]:
```suggestion
The module `model.py` contains classes that inherit from the
official Strava domain model in `strava_model.py`. This module, supports custom typing, unit conversion, (de-)serialization behavior, and support for
undocumented Strava features, .
```

not sure if this is what you mean just found this sentence a bit confusing as written.

**Comment by lwasser** [comment]
me again - @jsamoocha we could open an issue about this if you prefer? What do you want the model docs to look like? there are a bunch of ways to organize things so if i understand what you see as being "wrong" or not good enough I can fix them in a PR for docs only.

**Comment by jsamoocha** [comment]
> > * https://pypi.org/project/autodoc-pydantic/
> 
> No problem - if you want, i'll have a look at that and will see if i can get it to work more nicely

That would be awesome! I have no experience with Sphinx, so that'll save me from some headaches 🙏

**Comment by jsamoocha** [comment]
> ```
> response = []
> act = client.get_activities(limit=4)
> for anAct in act:
>     for aActResponse in anAct:
>         response.append(aActResponse)
> ```

This crashes on `master`, so I'd consider it an "un-breaking change" 😁. The inner loop iterates over a single `Activity`, which is not supported by the legacy entities. It seems like pydantic, by default, does support iteration over its model objects out of the box.

**Comment by jsamoocha** [comment]
> the response object is odd as it has both the bound client and then an athlete object that is full of "nones"

I agree that this is quirky, but it is the same behavior as on `master`.

Whenever you serialize to dict or json, the `bound_client` attribute is now excluded by default:

```
In [13]: acts[0]
Out[13]: Activity(bound_client=<stravalib.client.Client object at 0x106fa7250>, id=8447931995, achievement_count=0, athlete=MetaAthlete(bound_client=<stravalib.client.Client object at 0x106fa7250>, id=4233387), athlete_count=1, average_speed=0.0, average_watts=None, comment_count=0, commute=False, device_watts=None, distance=0.0, elapsed_time=997, elev_high=None, elev_low=None, end_latlng=None, external_id='garmin_ping_256953820006', flagged=False, gear_id=None, has_kudoed=False, hide_from_home=None, kilojoules=None, kudos_count=1, manual=False, map=Map(id='a8447931995', polyline=None, summary_polyline=''), max_speed=0.0, max_watts=None, moving_time=997, name='Morning Workout', photo_count=0, private=False, sport_type=<SportType.Workout: 'Workout'>, start_date=datetime.datetime(2023, 1, 25, 5, 47, 49, tzinfo=datetime.timezone.utc), start_date_local=datetime.datetime(2023, 1, 25, 6, 47, 49), start_latlng=None, timezone='(GMT+01:00) MET', total_elevation_gain=0.0, total_photo_count=0, trainer=True, type=<ActivityType.Workout: 'Workout'>, upload_id=9063513101, upload_id_str='9063513101', weighted_average_watts=None, workout_type=None, best_efforts=None, calories=None, description=None, device_name=None, embed_token=None, gear=None, laps=None, photos=None, segment_efforts=None, splits_metric=None, splits_standard=None, guid=None, utc_offset=3600.0, location_city=None, location_state=None, location_country='The Netherlands', start_latitude=None, start_longitude=None, pr_count=0, suffer_score=1, has_heartrate=True, average_heartrate=65.8, max_heartrate=88, average_cadence=53.0, average_temp=None, instagram_primary_photo=None, partner_logo_url=None, partner_brand_tag=None, from_accepted_tag=False, segment_leaderboard_opt_out=None)
In [14]: acts[0].dict()
Out[14]: 
{'id': 8447931995,  # <-- no bound client
 'achievement_count': 0,
...
```

It would be easy to use Strava's type hierarchy for different detail levels to prevent the mostly-filled-with-Nones nested entities, e.g. by rewriting our current `Athlete` entity as:

```python
class Athlete(DeprecatedSerializableMixin, BackwardCompatibilityMixin, BoundClientEntity):
    pass


class MetaAthlete(MetaAthlete, Athlete):  # <-- inherits from strava_model.MetaAthlete
    pass


class SummaryAthlete(SummaryAthlete, Athlete):
    pass

class DetailedAthlete(DetailedAthlete, Athlete):
    # field overrides from superclass for type extensions:
    clubs: Optional[List[Club]] = None
    bikes: Optional[List[Bike]] = None
    shoes: Optional[List[Shoe]] = None

    # Undocumented attributes:
    is_authenticated: Optional[bool] = None
    athlete_type: Optional[int] = None
    # ... etc
```

and then defining the `athlete` attribute of an `Activity` as:

```python
class Activity(DetailedActivity, BackwardCompatibilityMixin, DeprecatedSerializableMixin, BoundClientEntity):
    """
    Represents an activity (ride, run, etc.).
    """

    # field overrides from superclass for type extensions:
    athlete: Optional[MetaAthlete] = None  # <-- now using MetaAthlete per Strava API spec
    start_latlng: Optional[LatLon] = None
    # ...etc.
```

which would then result in:

```
In [6]: acts[0].athlete
Out[6]: MetaAthlete(bound_client=<stravalib.client.Client object at 0x106fa7250>, id=4233387) <-- No None's
...
In [10]: from stravalib.model import Athlete
In [11]: isinstance(acts[0].athlete, Athlete)
Out[11]: True  <-- and instance/type checks will still succeed
```

But:

```
In [15]: acts[0].athlete.firstname
Traceback (most recent call last):
  File "/Users/jsamoocha/.local/share/virtualenvs/stravalib_fork-PV0q2u3e/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3442, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-15-184a4e2651f5>", line 1, in <module>
    acts[0].athlete.firstname
  File "/Users/jsamoocha/dev/stravalib_fork/stravalib/model.py", line 145, in __getattribute__
    value = object.__getattribute__(self, attr)
AttributeError: 'MetaAthlete' object has no attribute 'firstname'
```

And this is a breaking change. I think we could (and should) implement this eventually because it is much cleaner. But to limit the confusion and other breaking changes to users, maybe postpone this?

**Comment by jsamoocha** [comment]
> What do you want the model docs to look like?

The main problem I saw was that our domain classes in the API reference had huge constructors with all the inherited fields from the generated model:

Model_-_Stravalib_Docs

Maybe using that autodoc plugin with default values already will resolve that?

**Comment by lwasser** [comment]
>Maybe using that autodoc plugin with default values already will resolve 
> that?

Ok @jsamoocha yup i have that fixed locally using that extension that you found for pydantic. How about I submit a PR to the pydantic_model branch as well with that docs fix.  i'm going to actually overhaul our entire API docs to make them cleaner. I had started this and never finished it :) 
Screen Shot 2023-01-29 at 9 00 16 AM
Screen Shot 2023-01-29 at 9 00 09 AM


> And this is a breaking change. I think we could (and should) implement this eventually because it is much > cleaner. But to limit the confusion and other breaking changes to users, maybe postpone this?

This makes sense. it would be nice to do as issues as well because then people can see the change as we discuss it (especially given it is breaking) and we can discuss options. 

Let's merge this PR into pydantic_model branch. Then we can take a bit of time to test it locally. 
In the meantime i'll clean up the docs. I'm also going to add the codespell hook to our pre-commit setup! i've been using it for some of my other repos and LOVE IT. It catches all sorts of small spelling errors in documentation!

**Review by lwasser — APPROVED** [review]
Approved - let's merge and continue to test on the pydantic branch! 🚀

**Comment by jsamoocha** [comment]
Ok, that API doc example already looks much better. Thanks for looking into it!

**Pull request merged**